File: lowestCommonAncestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # 思路：通过两个字典，一个字典记录二叉树每一个节点的父节点，另一个字典记录节点值和节点的对应关系
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        dic = {}
        node = {}
        self.reverse(root, node, dic)
        cur = p.val
        list1 = [cur]
        while cur in dic:
            cur = dic[cur]
            list1.append(cur)
        cur2 = q.val
        list2 = [cur2]
        while cur2 in dic:
            cur2 = dic[cur2]
            list2.append(cur2)
        for i in list1:
            if i in list2:
                return node[i]
        return root

    def reverse(self, root, node, dic):
        if not root:
            return 
        else:
            node[root.val] = root
            if root.left:
                dic[root.left.val] = root.val
            if root.right:
                dic[root.right.val] = root.val
            self.reverse(root.left, node, dic)
            self.reverse(root.right, node, dic)

File: lowestCommonAncestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        vector<TreeNode*> path1;
        TreeNode* cur = root;
        path1.push_back(cur);
        while(cur->val!=p->val){
                if(cur->val>p->val) cur = cur->left;
                else cur = cur->right;
                path1.push_back(cur);
        }
        cur = root;
        vector<TreeNode*> path2;
        path2.push_back(cur);
        while(cur->val!=q->val){
                if(cur->val>q->val) cur = cur->left;
                else cur = cur->right;
                path2.push_back(cur);
        }
        int i;
        int lowest;
        for(i = path1.size()-1;i>=0;i--){
            if(find(path2.begin(),path2.end(),path1[i])!= path2.end()){
                lowest = i;
                break;
            }
        }
        return path1[lowest];
    }
};
